### DISCUSSION TITLE
Linux System Theme not detected, missing dbus ReadOne method

### DISCUSSION THREAD

**Discussion by leigholiver** [ghostty-org/ghostty#3704] [Feature Requests, Ideas] [answered]
On my system (pop-os 22.04 LTS) with system theme and seperate light and dark themes, ghostty always defaults to light mode. Switches between light and dark mode are properly handled. 

In the logs this error is reported: 
```
error(gtk): unable to get current color scheme: GDBus.Error:org.freedesktop.DBus.Error.UnknownMethod: No such method “ReadOne”
```

The spec notes that the other functions are "[Deprecated, use `ReadOne` instead.](https://flatpak.github.io/xdg-desktop-portal/docs/doc-org.freedesktop.portal.Settings.html)" so using `ReadOne` is cerainly the correct path.

I've managed to fix this on my system by checking for the error and falling back to an implementation using the deprecated `Read`. The changes can be seen here https://github.com/leigholiver/ghostty/tree/leigholiver/dbus-deprecated

 I haven't seen any discussions around this issue specifically so wanted to open one before opening a PR. There are _similar_ issues around the system theme but the underlying issue is different.

**Comment by hpfr** [comment]
I ran into another GSettings issue in https://github.com/ghostty-org/ghostty/discussions/3771. As we read more prefs, the D-Bus boilerplate will multiply.

Since we already depend on GTK, should we be using [GSettings](https://docs.gtk.org/gio/class.Settings.html) (probably [`g_settings_new`](https://docs.gtk.org/gio/ctor.Settings.new.html) with `"org.freedesktop.appearance"`, [`g_settings_get_string`](https://docs.gtk.org/gio/method.Settings.get_string.html) with `"color-scheme"`), which should abstract away D-Bus and GVariant? 

Actually, for color scheme specifically, we can probably use [`adw_style_manager_get_color_scheme`](https://gnome.pages.gitlab.gnome.org/libadwaita/doc/1.6/method.StyleManager.get_color_scheme.html) when libadwaita is enabled (looks like it’s already set properly in `init`), falling back to GSettings when libadwaita is disabled.

**Reply by jcollie** [reply]
DBus is used because we need to react to changes in the color scheme for "live" switching between themes. AFAICS neither the GSettings or the Adwaita style manager allow subscribing to color changes. I think that the DBus method is more portable between DEs.

I think that merging your patch is a reasonable fix.

**Reply by hpfr** [reply]
> AFAICS neither the GSettings or the Adwaita style manager allow subscribing to color changes.

Looks like [Gio.Settings::changed](https://docs.gtk.org/gio/signal.Settings.changed.html) is a thing, but …

> I think that the DBus method is more portable between DEs.

… yeah, for the XDG color-scheme key specifically I think you’re right.

> I think that merging your patch is a reasonable fix.

Not my patch, but I agree :)

**Reply by hpfr** [reply]
Actually for libadwaita, [Adw.StyleManager](https://gnome.pages.gitlab.gnome.org/libadwaita/doc/1.6/class.StyleManager.html) says:

> Signals inherited from [GObject](https://docs.gtk.org/gobject/class.Object.html) (1)
[GObject::notify](https://docs.gtk.org/gobject/signal.Object.notify.html)
>
> The notify signal is emitted on an object when one of its properties has its value set through g_object_set_property(), g_object_set(), et al.

So using [GObject.Object::notify](https://docs.gtk.org/gobject/signal.Object.notify.html) on [Adw.StyleManager:dark](https://gnome.pages.gitlab.gnome.org/libadwaita/doc/1.6/property.StyleManager.dark.html) might work. But using D-Bus directly regardless of libadwaita is probably fine.

**Reply by jcollie** [reply]
We need a method that works without Adwaita (for now at least) so I think it's best to stick with the DBus method.

**Reply by leigholiver** [reply]
Thanks for the feedback both. The gsettings approach sounds like it might be cleaner, especially if other settings start to get involved (thinking about the large text discussion). I've done a quick search over the codebase and ReadOne doesnt appear to be used anywhere else. 
I need to check as im not sure, but if we're already depending on GTK, is gsettings any less portable?

**Reply by hpfr** [reply]
For `org.freedesktop.appearance` `color-scheme`, sticking with D-Bus is probably the right move to support other DEs with an XDG Desktop Portal implementation and Flatpak. GSettings still probably makes sense for the large text pref since that’s GNOME-specific. So your patch is using the right approach as it is, I think

**Comment by leigholiver** [comment]
I've opened an issue and PR for this

Issue: https://github.com/ghostty-org/ghostty/issues/4038
PR: https://github.com/ghostty-org/ghostty/pull/4035

**Comment by mitchellh** [comment, marked as answer]
This was fixed in 1.0.1
